File: harvard_CS50_compiuter_science/week_6_python/problem/readability.py

from cs50 import get_string


def main():
    # получаем юзеринпут в виде текста для анализа
    text = get_string('Text: ')
    # text = """One fish. Two fish. Red fish. Blue fish."""

    # в функциях вычисляем количество букв, слов и предложений в тексте
    letters = letters_count(text)
    words = words_count(text)
    sentiens = sentient_count(text)

    # получаем средние значения на 100 слов  букв и предложений
    letters_mid = (letters / words) * 100
    sentiens_mid = (sentiens / words) * 100

    # вычисляем сложность текста по формуле Коулмана-Ляу
    index = 0.0588 * letters_mid - 0.296 * sentiens_mid - 15.8

    # в зависимости от результата выводим сложность текста
    if index < 1:
        print('Before Grade 1')
    elif index > 13:
        print('Grade 16+')
    else:
        print(f'Grade {index:.0f}')


def letters_count(text):

    letters = 0
    for i in text:
        if i.isalpha():
            letters += 1

    return letters


def words_count(text):

    # ставим минимальное значение 1, для учета последнего слова
    words = 1
    for i in text:
        if i == ' ':
            words += 1

    return words


def sentient_count(text):

    sentiens = 0
    for i in text:
        if i == '.' or i == '!' or i == '?':
            sentiens += 1

    return sentiens


if __name__ == '__main__':
    main()

File: CS50x/week_2_array/problem/readability/readability.c
#include <ctype.h>
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

int letters_count(string text);
int words_count(string text);
int sentient_count(string text);

int main(void) 
{
    string text = get_string("Text: ");
    int letters = letters_count(text);
    int words = words_count(text);
    int sentiens = sentient_count(text);

    float letters_mid = ((float) letters / words) * 100;
    float sentiens_mid = ((float) sentiens / words) * 100;

    float index = 0.0588 * letters_mid - 0.296 * sentiens_mid - 15.8;

    if (index < 1)
    {
        printf("Before Grade 1\n");
    }

    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }

    else
    {
        printf("Grade %i\n", (int) round(index));
    }
    
}

int letters_count(string text)
{    
    int letter_count = 0;
    char c = 0;

    for (int i = 0; i < strlen(text); i++)
    {
        c = text[i];
        if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z'))
        {
            letter_count = letter_count + 1;
        }

    }

    return letter_count;
}

int words_count(string text)
{
    int counter = 0;
    char c = 0;

    for (int i = 0; i < strlen(text); i++)		
    {
        c = text[i];
        if ((c == ' ') || (c == '.') || (c == '!') || (c == ','))
        {
            counter = counter + 1;
            if (text[i + 1] == ' ')
            {
                i = i + 1;
            }
        }
        
    }

    return counter;
}

int sentient_count(string text)
{
    int counter = 0;
    char c = 0;

    for (int i = 0; i < strlen(text); i++)		
    {
        c = text[i];
        if ((c == '.') || (c == '!') || (c == '?'))
        {
            counter = counter + 1;
        }
    }
    return counter;
}

